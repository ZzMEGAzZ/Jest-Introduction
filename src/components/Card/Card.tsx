import React from 'react';

interface CardProps {
  children: React.ReactNode;
  title?: string;
  variant?: 'default' | 'outlined' | 'elevated';
  className?: string;
  onClick?: () => void;
}

export const Card: React.FC<CardProps> = ({
  children,
  title,
  variant = 'default',
  className = '',
  onClick
}) => {
  const baseStyles = 'rounded-lg p-4';
  const variantStyles = {
    default: 'bg-white shadow',
    outlined: 'border border-gray-200',
    elevated: 'bg-white shadow-lg'
  };

  return (
    <div
      className={`${baseStyles} ${variantStyles[variant]} ${className}`}
      onClick={onClick}
      data-testid="card"
      role={onClick ? 'button' : 'article'}
    >
      {title && (
        <h1 className="text-lg font-semibold mb-2" data-testid="card-title">
          {title}
        </h1>
      )}
      <div data-testid="card-content">
        {children}
      </div>
    </div>
  );
};