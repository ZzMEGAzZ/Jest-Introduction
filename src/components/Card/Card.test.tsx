import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Card } from "./Card";

describe("Card", () => {
  test("renders children content", () => {
    render(
      <Card>
        <p>Test content</p>
      </Card>
    );
    expect(screen.getByText("Test content")).toBeInTheDocument();
  });

  test("renders title when provided", () => {
    render(
      <Card title="Test Title">
        <p>Content</p>
      </Card>
    );
    expect(screen.getByTestId("card-title")).toHaveTextContent("Test Title");
  });

  test("applies different variant styles", () => {
    const { rerender } = render(
      <Card variant="outlined">
        <p>Content</p>
      </Card>
    );
    expect(screen.getByTestId("card")).toHaveClass("border", "border-gray-200");

    rerender(
      <Card variant="elevated">
        <p>Content</p>
      </Card>
    );
    expect(screen.getByTestId("card")).toHaveClass("shadow-lg");
  });

  test("handles click events", () => {
    const handleClick = jest.fn();
    render(
      <Card onClick={handleClick}>
        <p>Clickable content</p>
      </Card>
    );

    fireEvent.click(screen.getByTestId("card"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("has correct role when clickable", () => {
    const { rerender } = render(
      <Card onClick={() => {}}>
        <p>Content</p>
      </Card>
    );
    expect(screen.getByTestId("card")).toHaveAttribute("role", "button");

    rerender(
      <Card>
        <p>Content</p>
      </Card>
    );
    expect(screen.getByTestId("card")).toHaveAttribute("role", "article");
  });

  test("applies custom className", () => {
    render(
      <Card className="custom-class">
        <p>Content</p>
      </Card>
    );
    expect(screen.getByTestId("card")).toHaveClass("custom-class");
  });

  // Integration test with nested components
  test("works with nested components", () => {
    render(
      <Card title="Profile Card">
        <div>
          <h4>John Doe</h4>
          <p>Developer</p>
          <button>Contact</button>
        </div>
      </Card>
    );

    expect(screen.getByText("Profile Card")).toBeInTheDocument();
    expect(screen.getByText("John Doe")).toBeInTheDocument();
    expect(screen.getByText("Developer")).toBeInTheDocument();
    expect(screen.getByText("Contact")).toBeInTheDocument();
  });
});

describe("Card", () => {
  test("matches snapshot", () => {
    const { container } = render(
      <Card title="Test Card">
        <p>Test content</p>
      </Card>
    );
    expect(container).toMatchSnapshot();
  });

  // Test with different variants
  test("matches snapshot with different variants", () => {
    const { container: defaultCard } = render(
      <Card variant="default">
        <p>Default variant</p>
      </Card>
    );
    expect(defaultCard).toMatchSnapshot("default variant");

    const { container: outlinedCard } = render(
      <Card variant="outlined">
        <p>Outlined variant</p>
      </Card>
    );
    expect(outlinedCard).toMatchSnapshot("outlined variant");
  });
});
