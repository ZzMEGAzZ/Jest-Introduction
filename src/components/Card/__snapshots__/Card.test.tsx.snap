// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches snapshot 1`] = `
<div>
  <div
    class="rounded-lg p-4 bg-white shadow "
    data-testid="card"
    role="article"
  >
    <h3
      class="text-lg font-semibold mb-2"
      data-testid="card-title"
    >
      Test Card
    </h3>
    <div
      data-testid="card-content"
    >
      <p>
        Test content
      </p>
    </div>
  </div>
</div>
`;

exports[`Card matches snapshot with different variants: default variant 1`] = `
<div>
  <div
    class="rounded-lg p-4 bg-white shadow "
    data-testid="card"
    role="article"
  >
    <div
      data-testid="card-content"
    >
      <p>
        Default variant
      </p>
    </div>
  </div>
</div>
`;

exports[`Card matches snapshot with different variants: outlined variant 1`] = `
<div>
  <div
    class="rounded-lg p-4 border border-gray-200 "
    data-testid="card"
    role="article"
  >
    <div
      data-testid="card-content"
    >
      <p>
        Outlined variant
      </p>
    </div>
  </div>
</div>
`;
